srtcore/buffer.cpp

```diff
@@ -523,7 +523,19 @@ int CSndBuffer::readData(const int offset, srt::CPacket& w_packet, steady_clock:
     // XXX Suboptimal procedure to keep the blocks identifiable
     // by sequence number. Consider using some circular buffer.
     for (int i = 0; i < offset; ++i)
+    {
+        if (p == NULL || p == m_pLastBlock)
+        {
+            LOGC(qslog.Error, log << "CSndBuffer::readData: offset " << offset << " too large!");
+            return -2;
+        }
         p = p->m_pNext;
+    }
+    if (p == NULL || p == m_pLastBlock)
+    {
+        LOGC(qslog.Error, log << "CSndBuffer::readData: offset " << offset << " too large!");
+        return -2;
+    }
```

Comment:
It might be a good idea to return 0 instead of `-2`. 🤔 
There is already a check for `0` as a return. It would be an indication that nothing was extracted. While `-1` indicates that there is data, but it must be dropped.

Reply:
I return `-2` since there is a check:
```
        const int payload = m_pSndBuffer->readData(offset, (w_packet), (w_origintime), (msglen));
        SRT_ASSERT(payload != 0);
```
So I just remove this assert?

Reply:
Yes, I would remove this assertion. Maybe an error log message here would make more sense (moving it from `readData`).
```c++
else if (payload == 0)
        continue;
```

Previously returning 0 was something impossible. It meant that a packet was found in the buffer, but it has zero length.

With the additional check you are adding, returning 0 would still mean something unexpected: request for retransmission of a packet never sent (from the future).
`offset` is the offset from the last acknowledged sequence number. If retransmission of a packet that was never sent is requested, the receiver requesting retransmission is doing something weird.

Reply:
```c++
else if (payload == 0)
{
    LOGC(qslog.Error, log << "IPE/EPE: packLostData: couldn't find a packet with seqno " << w_packet.m_iSeqNo << "(offset " << offset << ").");
    continue;
}
```